### PR TITLE
BUG: Ensure einsum uses chunking (now that nditer doesn't)

### PR DESCRIPTION
**Pull request by seberg**
The nditer refactor enabled the `GROWINNER` for reductions, this was not the case before (I am not sure about cases where the reduction is in an outer axis, but I don't think so).

Either way, as the test says, chunking always improves the precision of large sums if their mean is nonzero and sklearn noticed lower precision.

We could only remove growinner if there is a reduction, but it seems like a 1% performance hit for the simplest (non-trivial) case where GROWINNER had done something before.

---

I suspect this is a pragmatic thing (little downside, better precision for large arrays).  An even better thing would be pairwise summation, within einsum I suppose.

I.e. the reason for the chunking used to be that the iterator didn't manage to optimize this (an explicit code comment).  But now, it became a choice to help the sum precision a bit...

See also https://github.com/scikit-learn/scikit-learn/issues/30509, ping @lesteve

### CODE REVIEW

**Comment by lesteve** [comment]
I can confirm that this PR fixes https://github.com/scikit-learn/scikit-learn/issues/30509.

File: numpy/_core/tests/test_einsum.py [hunk 1]
```diff
@@ -1282,3 +1282,29 @@ def test_overlap():
...
+    (same effect as pairwise summation but much less).
+    Without adding pairwise summation to the inner-loop it seems best to just
+    not use GROWINNER, a quick tests suggest that is maybe 1% slowdown for
+    the simplest `einsum("i,i->i", x, x)` case.
+
+    (It is not clear that we should guarantee precision to this extend.)
+    """
+    num = 100000000
+    value = 1.00000000002
+    res = np.einsum("i->", np.full(num, value)) / num
```
Comment (mhvk) [inline numpy/_core/tests/test_einsum.py:1308]: I'd tend to use `float32` and fewer elements, just to keep the test fast (now 147 ms on my machine).

File: numpy/_core/src/multiarray/einsum.c.src [hunk 2]
```diff
@@ -1035,7 +1035,6 @@ PyArray_EinsteinSum(char *subscripts, npy_intp nop,
     iter_flags = NPY_ITER_EXTERNAL_LOOP|
             NPY_ITER_BUFFERED|
             NPY_ITER_DELAY_BUFALLOC|
-            NPY_ITER_GROWINNER|
```
Comment (mhvk) [inline numpy/_core/src/multiarray/einsum.c.src:1038]: How about adding a comment here? Partially as a reminder that a pair-wise sum might be a good idea...

**Review by mhvk — COMMENTED** [review]
Seems a sensible solution for now. Two small comments inline.

File: numpy/_core/tests/test_einsum.py (hunk 1 above)
Comment (seberg) [reply numpy/_core/tests/test_einsum.py:1308]: Doesn't work :).  Einsum uses double for the accumulation, so for float32, the results are actually better without chunking...

But, let me use broadcasting and make the number smaller, certainly doesn't need this much.

File: numpy/_core/src/multiarray/einsum.c.src (hunk 2 above)
Comment (seberg) [reply numpy/_core/src/multiarray/einsum.c.src:1038]: Yeah, didn't think of the "let's do pairwise" spin, so wasn't sure.

**Review by mhvk — APPROVED** [review]
Super!

**Comment by lesteve** [comment]
Great, thanks a lot for the investigation and the fix!

Maybe one day I will manage to understand this tricky floating point things, maybe one day :sweat_smile:.

**Comment by seberg** [comment]
> Maybe one day I will manage to understand these tricky floating point things

Conceptually, (in decimals) it comes down to:
Say `10. + 1.0001 == 1` in floating point math but is exact for smaller values (not true, but...).
Then `sum([1.0001] * 10)` might give you `10.001` exactly, but `sum([1.0001] * 20)` will be the reach `10.001` and after that ignore the `0.0001` part and return `20.001` rather than `20.002`.

But if you chunk it up into batches of 10 `sum([1.0001] * 10) + sum([1.0001] * 10)` you get `10.001 + 10.001` which is precise:  It is more precise to add similar values and chunking/ or pairwise summation `((a0 + a1) + (a2 + a3)) + ...` tends to add more similar values.
